File: sigma-tree/src/chain/contract.rs
//! Ergo contract

use super::address::Address;
use crate::{serialization::SerializationError, ErgoTree};

/// High-level wrapper for ErgoTree
#[derive(PartialEq, Eq, Debug, Clone)]
pub struct Contract {
    ergo_tree: ErgoTree,
}

impl Contract {
    /// create new contract from ErgoTree
    pub fn new(ergo_tree: ErgoTree) -> Contract {
        Contract { ergo_tree }
    }

    /// create new contract that allow spending for a given Address
    pub fn pay_to_address(address: Address) -> Result<Contract, SerializationError> {
        Ok(Contract::new(address.script()?))
    }

    /// get ErgoTree for this contract
    pub fn get_ergo_tree(&self) -> ErgoTree {
        self.ergo_tree.clone()
    }
}

File: bindings/ergo-lib-ios/Sources/ErgoLib/Contract.swift
import Foundation
import ErgoLibC

/// Defines the contract(script) that will be guarding box contents
class Contract {
    internal var pointer: ContractPtr
    
    internal init(withRawPtr ptr: ContractPtr) {
        self.pointer = ptr
    }
    
    /// Create new contract from ErgoTree
    init(fromErgoTree: ErgoTree) {
        var ptr:  ContractPtr?
        ergo_lib_contract_new(fromErgoTree.pointer, &ptr)
        self.pointer = ptr!
    }
    
    /// Compiles a contract from ErgoScript source code
    init(compileFromString: String) throws {
        var ptr: ContractPtr?
        let error = compileFromString.withCString { cs in
            ergo_lib_contract_compile(cs, &ptr)
        }
        try checkError(error)
        self.pointer = ptr!
    }
    
    /// Create new contract that allow spending of the guarded box by a given recipient ([`Address`])
    init(payToAddress: Address) throws {
        var ptr: ContractPtr?
        let error = ergo_lib_contract_pay_to_address(payToAddress.pointer, &ptr)
        try checkError(error)
        self.pointer = ptr!
    }
    
    /// Get the ErgoTree of the contract
    func getErgoTree() -> ErgoTree {
        var boxIdPtr: ErgoTreePtr?
        ergo_lib_contract_ergo_tree(self.pointer, &boxIdPtr)
        return ErgoTree(withRawPointer: boxIdPtr!)
    }
        
    deinit {
        ergo_lib_contract_delete(self.pointer)
    }
}

extension Contract: Equatable {
    static func ==(lhs: Contract, rhs: Contract) -> Bool {
        ergo_lib_contract_eq(lhs.pointer, rhs.pointer)
    }
}
